### PR TITLE
MGRDatabaseDiscoveryType add check data source does it exist mgr group.

### LINKED ISSUE #14646: Optimize Database discovery MGR prompt information.
### Which version of ShardingSphere did you use?

`master`

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

`ShardingSphere-Proxy`

### Expected behavior

Remind which dataSource not joined MGR cluster.

### Actual behavior

```
Exception in thread "main" java.lang.NullPointerException
	at org.apache.shardingsphere.dbdiscovery.mgr.MGRDatabaseDiscoveryType.findMemberDataSourceURLs(MGRDatabaseDiscoveryType.java:189)
	at org.apache.shardingsphere.dbdiscovery.mgr.MGRDatabaseDiscoveryType.updateMemberState(MGRDatabaseDiscoveryType.java:178)
	at org.apache.shardingsphere.dbdiscovery.rule.DatabaseDiscoveryRule.findMasterSlaveRelation(DatabaseDiscoveryRule.java:123)
	at org.apache.shardingsphere.dbdiscovery.rule.DatabaseDiscoveryRule.<init>(DatabaseDiscoveryRule.java:85)
	at org.apache.shardingsphere.dbdiscovery.rule.DatabaseDiscoveryRule.<init>(DatabaseDiscoveryRule.java:74)
	at org.apache.shardingsphere.dbdiscovery.rule.builder.DatabaseDiscoveryRuleBuilder.build(DatabaseDiscoveryRuleBuilder.java:46)
	at org.apache.shardingsphere.dbdiscovery.rule.builder.DatabaseDiscoveryRuleBuilder.build(DatabaseDiscoveryRuleBuilder.java:36)
	at org.apache.shardingsphere.infra.rule.builder.schema.SchemaRulesBuilder.buildRules(SchemaRulesBuilder.java:85)
	at org.apache.shardingsphere.infra.rule.builder.schema.SchemaRulesBuilder.buildRules(SchemaRulesBuilder.java:70)
	at org.apache.shardingsphere.mode.manager.cluster.ClusterContextManagerBuilder.beforeBuildContextManager(ClusterContextManagerBuilder.java:103)
	at org.apache.shardingsphere.mode.manager.cluster.ClusterContextManagerBuilder.build(ClusterContextManagerBuilder.java:84)
	at org.apache.shardingsphere.proxy.initializer.BootstrapInitializer.createContextManager(BootstrapInitializer.java:84)
	at org.apache.shardingsphere.proxy.initializer.BootstrapInitializer.init(BootstrapInitializer.java:71)
	at org.apache.shardingsphere.proxy.Bootstrap.main(Bootstrap.java:47)

```

### PR DESCRIPTION
**Pull request by zjcnb**
Fixes #14646

### CODE REVIEW

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/rule/DatabaseDiscoveryRule.java [hunk 1]
```diff
@@ -81,6 +81,7 @@ public DatabaseDiscoveryRule(final String schemaName, final Map<String, DataSour
     
     private DatabaseDiscoveryRule(final String schemaName, final Map<String, DataSource> dataSourceMap, final Collection<DatabaseDiscoveryDataSourceRuleConfiguration> dataSourceRuleConfigs,
                                   final Map<String, DatabaseDiscoveryHeartBeatConfiguration> heartBeatConfig, final Map<String, DatabaseDiscoveryType> discoveryTypes) {
+        Preconditions.checkState(!dataSourceMap.isEmpty(), "Data source can not be is empty.");
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/rule/DatabaseDiscoveryRule.java:84]: Data source can not be empty.

Comment (zjcnb) [reply shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/rule/DatabaseDiscoveryRule.java:84]: done.

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java [hunk 2]
```diff
@@ -114,6 +116,27 @@ private void checkIsSinglePrimaryMode(final Statement statement) throws SQLExcep
         }
     }
     
+    private void checkDataSourceInReplicationGroup(final Statement statement, final Map<String, DataSource> dataSourceMap) throws SQLException {
+        List<String> memberDataSourceURLs = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (zjcnb) [reply shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java]: Ok, i will adjust.

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java [hunk 3]
```diff
@@ -114,6 +115,31 @@ private void checkIsSinglePrimaryMode(final Statement statement) throws SQLExcep
...
+        List<String> memberDataSourceURLs = new LinkedList<>();
+        try (ResultSet resultSet = statement.executeQuery(MEMBER_LIST)) {
+            while (resultSet.next()) {
+                memberDataSourceURLs.add(String.format("%s:%s", resultSet.getString("MEMBER_HOST"), resultSet.getString("MEMBER_PORT")));
+            }
+        }
+        checkDataSourcesExistGroupMember(dataSourceMap, memberDataSourceURLs);
+    }
+    
+    private void checkDataSourcesExistGroupMember(final Map<String, DataSource> dataSourceMap, final List<String> memberDataSourceURLs) throws SQLException {
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java:128]: `List<String> memberDataSourceURLs` can be `Collection<String> memberDataSourceURLs`

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java [hunk 4]
```diff
@@ -114,6 +115,31 @@ private void checkIsSinglePrimaryMode(final Statement statement) throws SQLExcep
         }
     }
     
+    private void checkDataSourceInReplicationGroup(final Statement statement, final Map<String, DataSource> dataSourceMap) throws SQLException {
+        List<String> memberDataSourceURLs = new LinkedList<>();
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java:119]: `Collection<String> memberDataSourceURLs` can be `List<String> memberDataSourceURLs`

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java [hunk 5]
```diff
@@ -114,6 +115,31 @@ private void checkIsSinglePrimaryMode(final Statement statement) throws SQLExcep
...
+            boolean exist = false;
+            for (String each : memberDataSourceURLs) {
+                if (entry.getValue().getConnection().getMetaData().getURL().contains(each)) {
+                    exist = true;
+                    break;
+                }
+            }
+            if (!exist) {
+                throw new ShardingSphereConfigurationException("%s is not MGR replication group member", entry.getKey());
+            }
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java:139]: It is better to keep one loop in single method

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java [hunk 6]
```diff
@@ -114,6 +115,31 @@ private void checkIsSinglePrimaryMode(final Statement statement) throws SQLExcep
...
+                memberDataSourceURLs.add(String.format("%s:%s", resultSet.getString("MEMBER_HOST"), resultSet.getString("MEMBER_PORT")));
+            }
+        }
+        for (Entry<String, DataSource> entry : dataSourceMap.entrySet()) {
+            checkDataSourceExistGroupMember(entry.getKey(), entry.getValue(), memberDataSourceURLs);
+        }
+    }
+    
+    private void checkDataSourceExistGroupMember(final String datasourceName, final DataSource dataSource, final Collection<String> memberDataSourceURLs) throws SQLException {
+        boolean exist = false;
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java:131]: `exist` is a verb, how about rename it as `isExisted`?

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java [hunk 7]
```diff
@@ -114,6 +115,31 @@ private void checkIsSinglePrimaryMode(final Statement statement) throws SQLExcep
...
+            while (resultSet.next()) {
+                memberDataSourceURLs.add(String.format("%s:%s", resultSet.getString("MEMBER_HOST"), resultSet.getString("MEMBER_PORT")));
+            }
+        }
+        for (Entry<String, DataSource> entry : dataSourceMap.entrySet()) {
+            checkDataSourceExistGroupMember(entry.getKey(), entry.getValue(), memberDataSourceURLs);
+        }
+    }
+    
+    private void checkDataSourceExistGroupMember(final String datasourceName, final DataSource dataSource, final Collection<String> memberDataSourceURLs) throws SQLException {
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java:130]: It is better to rename `checkDataSourceExistGroupMember` as `checkDataSourceExistedWithGroupMember`

Comment (zjcnb) [reply shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java]: Ok.

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java (hunk 6 above)
Comment (zjcnb) [reply shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-provider/shardingsphere-db-discovery-mgr/src/main/java/org/apache/shardingsphere/dbdiscovery/mgr/MGRDatabaseDiscoveryType.java]: Done.

**Review by terrymanu — APPROVED** [review]

**Pull request merged**